DOCS-1197: Review modular resource docs (#2106)
viamrobotics/docs: docs/modular-resources/configure.md

    "components",
    "services",
  ]
description: "Add a modular resource to your robot by configuring it."
no_list: true
aliases:
  - "/program/extend/modular-resources/configure/"
---

To use a [modular resource](/modular-resources/) on your robot, you need to:

1. Configure a {{< glossary_tooltip term_id="module" text="module" >}}, either one [from the registry](#add-a-module-from-the-viam-registry) or a [local module](#local-modules).
   This makes the modular resource available to the robot.

> 1. Then configure the modular resource itself.

Reviewer: You can add a modular resource directly without needing to do anything with the module that provides it. Most users will add the modular resource and not bother with any module-specific conf, in fact.

Conceptually - I agree - it makes sense to add the module, then add the modular resource it provides. But in practice, eng has designed the process so that the MRs can be added directly without needing awareness of its parent module.

To see in action:

App: Config > Create Component > `camera / realsense` > Add module > name it > Create > Poof module also added with default settings. No module conf required to start using realsense MR.

Author: Hmm. It is very streamlined but as you say you do click "Add module" before configuring the resource itself. That button configures the module. 

The current version of this intro section talks all about modules and says nothing about configuring the resource itself, so it sounds like we need to change that in some way, just need to figure out in which way, does that sound right?

Reviewer: It does. Perhaps the heavy usage of 'Module' on this page should be 'modular resource' in more places, I could see that!

Reviewer: Although, I guess I'd like to see the Add module button read "Add modular X" (component or service), since that's actually what is taking place (we came here from X tab, not Modules tab. We are adding to X tab, not modules tab -- although the providing module is also added to modules behind the scenes. Ex: We see a new camera component appear when we click, not the intel-realsense module appear).

The text itself "Add module" seems like a stretch to consider as a configuration step. Configuration seems to me to be more update management, and the upcoming [module environment variable support](https://github.com/viamrobotics/docs/pull/2102). Those steps configure the _module specifically_, so I think there is some cause to differentiate between configuring a module and its modular resource.

The real fix is better button text though.

Author: I thought "Add module" is correct because that button adds the module on the modules tab, and then you continue configuring the modular resource in place?

Reviewer: I think you've really helped to clarify things with the latest edit. "Underlying module" is the language I've been looking for all this time, ha! Your callout about the module being added at the same time as the modular resource I think is also a great step forward, but have added a slight language suggestion as well. Thanks for thinking through this with me!

    "components",
    "services",
  ]
description: "Add a modular resource to your robot by configuring it."
no_list: true
aliases:
  - "/program/extend/modular-resources/configure/"
---

To use a [modular resource](/modular-resources/) on your robot, you need to:

1. Add a {{< glossary_tooltip term_id="module" text="module" >}}, either one [from the registry](#add-a-modular-resource-from-the-viam-registry) or a [local module](#local-modules).
   This makes the modular resource available to the robot.
1. Then add the modular resource itself.

> These two steps automatically happen at the same time when you add a modular resource from the registry.

Reviewer: What do you think of this rephrase:

```suggestion
When you add a modular resource from the registry, the underlying module that provides it is automatically added at the same time.
To add a modular resource from a local module, you must add the module first.
```

- "is automatically added" rather than "time" so we explicitly note the (absence of) user agency: no action required for the underlying module when from registry.
- Bit more explicit on what to do for each path to adding a modular resource, especially since we just introduced the two just above. Now we state how to initially interact with modules for both as well.

Author: good points. I like it.

   {{<imgproc src="modular-resources/configure/add-component-by-name.png" resize="400x" declaredimensions=true alt="The add a component modal showing results for the intel realsense module when searching by the name 'realsense'">}}

> 1. After selecting the modular component, click the **Add module** button, enter a name for your modular component, and click **Create** to add it to your robot's component config.

Reviewer: I would only use the shorthand when it matches the UI.

```suggestion
1. After selecting the modular component, click the **Add module** button, enter a name for your modular component, and click **Create** to add it to your robot's component configuration.
```

   {{<imgproc src="modular-resources/configure/add-component-screen.png" resize="400x" declaredimensions=true alt="The add a component modal showing the intel realsense module pane, with the 'Add module' button shown">}}

   Be sure the modular component you select supports the [platform](/manage/cli/#using-the---platform-argument) you intend to use it with, such as `linux arm64`.
   You can see which platforms the module supports at bottom of the module information screen before you add it.

When you add a module from the Viam registry, the custom modular component it provides appears under the **Components** subtab like any other component.

> You can also find [the underlying module](#edit-the-configuration-of-a-module-from-the-viam-registry) listed as **Deployed** under the **Modules** subtab.

Reviewer: Underlying!! Yes, this is the right word, perfect. Thanks for finding it!

   {{<imgproc src="modular-resources/configure/add-service-by-name.png" resize="400x" declaredimensions=true alt="The add a component modal showing results for the mlmodelservice triton module when searching by the name 'triton'">}}

> 1. After selecting the modular service, click the **Add module** button, enter a name for your modular service, and click **Create** to add it to your robot's service config.

Reviewer: ```suggestion
1. After selecting the modular service, click the **Add module** button, enter a name for your modular service, and click **Create** to add it to your robot's service configuration.
```

You can also find [the module itself](#edit-the-configuration-of-a-module-from-the-viam-registry) listed as **Deployed** under the **Modules** subtab.

If the module requires you to configure specific **Attributes**, click the **URL** link in the [module's configuration pane](#edit-the-configuration-of-a-module-from-the-viam-registry) to view the specific attribute requirements on the module's GitHub page.

To delete a module added from the Viam registry, click the trash can icon in the upper-right corner of the module configuration pane in the **Services** tab.
Deleting a module _does not_ delete any configured modular resources it provides.

### Add additional modular resources from a registry module

Once you have [added a module](#add-a-modular-resource-from-the-viam-registry) from the Viam registry, you can add any number of the modular resources it provides to your robot by adding new components or services configured with your modular resource's [model](/modular-resources/key-concepts/#models).

Follow the same steps as when you added the first modular resource, clicking **Create component** or **Create service** as applicable.

> You will be prompted to click **Add module** again while configuring the resource, though no duplicate module will be added to the `modules` section of the config.

Reviewer: ```suggestion
You will be prompted to click **Add module** again while configuring the resource, though no duplicate module will be added to the `modules` section of the configuration.
```

   This example shows the configuration for adding a [CSI camera](/modular-resources/examples/csi/) as a local module.

> ### Edit local module configuration

Reviewer: ```suggestion
### Edit the configuration of a local module
```

> Once you have added a modular resource to your robot, you can view and configure the module itself from the **Modules** subtab:

Reviewer: ```suggestion
Once you have added a modular resource to your robot, you can view and edit the underlying module from the **Modules** subtab:
```

I like "underlying"! 👍
